File: src/lab_metrics/store.py
"""نوشتن metricهای دقیقه‌ای، gauge و event در Redis."""

from __future__ import annotations

import json
import time
from typing import Any

from redis import Redis

from .labels import metric_field


LATENCY_BUCKETS_MS = (100, 250, 500, 1000, 2500, 5000, 10000, 30000, 60000)


class MetricStore:
    def __init__(
        self,
        redis: Redis,
        *,
        prefix: str = "lab:metrics:v1",
        retention_seconds: int = 172800,
        event_retention_seconds: int = 604800,
        event_limit: int = 200,
    ) -> None:
        self.redis = redis
        self.prefix = prefix
        self.retention_seconds = max(3600, retention_seconds)
        self.event_retention_seconds = max(3600, event_retention_seconds)
        self.event_limit = max(20, event_limit)

    def increment(
        self,
        name: str,
        amount: float = 1,
        labels: dict[str, object] | None = None,
        *,
        timestamp: float | None = None,
    ) -> None:
        key = self.minute_key(timestamp)
        pipe = self.redis.pipeline(transaction=False)
        pipe.hincrbyfloat(key, metric_field(name, labels), amount)
        pipe.expire(key, self.retention_seconds)
        _execute(pipe)

    def timing(
        self,
        name: str,
        elapsed_ms: float,
        labels: dict[str, object] | None = None,
        *,
        timestamp: float | None = None,
    ) -> None:
        key = self.minute_key(timestamp)
        pipe = self.redis.pipeline(transaction=False)
        pipe.hincrbyfloat(key, metric_field(f"{name}.count", labels), 1)
        pipe.hincrbyfloat(key, metric_field(f"{name}.sum_ms", labels), elapsed_ms)
        for boundary in LATENCY_BUCKETS_MS:
            if elapsed_ms <= boundary:
                pipe.hincrbyfloat(
                    key,
                    metric_field(f"{name}.le_{boundary}", labels),
                    1,
                )
        pipe.hincrbyfloat(key, metric_field(f"{name}.le_inf", labels), 1)
        pipe.expire(key, self.retention_seconds)
        _execute(pipe)

    def gauge(
        self,
        name: str,
        value: object,
        labels: dict[str, object] | None = None,
    ) -> None:
        key = f"{self.prefix}:gauges"
        pipe = self.redis.pipeline(transaction=False)
        pipe.hset(key, metric_field(name, labels), str(value))
        pipe.expire(key, self.retention_seconds)
        _execute(pipe)

    def event(self, category: str, event: str, **fields: Any) -> None:
        payload = {
            "timestamp": int(time.time() * 1000),
            "category": category,
            "event": event,
            **{key: value for key, value in fields.items() if value is not None},
        }
        key = f"{self.prefix}:events"
        pipe = self.redis.pipeline(transaction=False)
        pipe.lpush(key, json.dumps(payload, ensure_ascii=False, separators=(",", ":")))
        pipe.ltrim(key, 0, self.event_limit - 1)
        pipe.expire(key, self.event_retention_seconds)
        _execute(pipe)

    def begin_request(self, request_id: str, *, timeout_seconds: int = 300) -> None:
        key = f"{self.prefix}:inflight"
        pipe = self.redis.pipeline(transaction=False)
        pipe.zadd(key, {request_id: time.time() + timeout_seconds})
        pipe.expire(key, timeout_seconds * 2)
        _execute(pipe)

    def end_request(self, request_id: str) -> None:
        try:
            self.redis.zrem(f"{self.prefix}:inflight", request_id)
        except Exception:
            pass

    def inflight(self) -> int:
        key = f"{self.prefix}:inflight"
        pipe = self.redis.pipeline(transaction=False)
        pipe.zremrangebyscore(key, "-inf", time.time())
        pipe.zcard(key)
        try:
            _, count = pipe.execute()
            return int(count)
        except Exception:
            return 0

    def minute_key(self, timestamp: float | None = None) -> str:
        minute = int((timestamp or time.time()) // 60)
        return f"{self.prefix}:minute:{minute}"


def _execute(pipe) -> None:
    try:
        pipe.execute()
    except Exception:
        # metric هیچ‌وقت نباید business request را fail کند.
        pass

File: go/metrics/store.go
package metrics

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"time"

	"github.com/redis/go-redis/v9"
)

var latencyBuckets = []float64{100, 250, 500, 1000, 2500, 5000, 10000, 30000, 60000}

type Store struct {
	Redis                     *redis.Client
	Prefix                    string
	Retention, EventRetention time.Duration
	EventLimit                int
}

func FromEnv() (*Store, error) {
	endpoint := os.Getenv("REDIS_URL")
	if endpoint == "" {
		endpoint = "redis://localhost:6379/0"
	}
	options, err := redis.ParseURL(endpoint)
	if err != nil {
		return nil, err
	}
	return &Store{Redis: redis.NewClient(options), Prefix: valueOr(os.Getenv("METRICS_NAMESPACE"), "lab:metrics:v1"), Retention: time.Duration(intEnv("METRICS_RETENTION_SECONDS", 172800)) * time.Second, EventRetention: time.Duration(intEnv("METRICS_EVENT_RETENTION_SECONDS", 604800)) * time.Second, EventLimit: intEnv("METRICS_EVENT_LIMIT", 200)}, nil
}
func (s *Store) Close() error { return s.Redis.Close() }
func (s *Store) Increment(ctx context.Context, name string, amount float64, labels map[string]any) {
	s.write(ctx, func(pipe redis.Pipeliner, key string) { pipe.HIncrByFloat(ctx, key, Field(name, labels), amount) })
}
func (s *Store) Timing(ctx context.Context, name string, elapsed float64, labels map[string]any) {
	s.write(ctx, func(pipe redis.Pipeliner, key string) {
		pipe.HIncrByFloat(ctx, key, Field(name+".count", labels), 1)
		pipe.HIncrByFloat(ctx, key, Field(name+".sum_ms", labels), elapsed)
		for _, boundary := range latencyBuckets {
			if elapsed <= boundary {
				pipe.HIncrByFloat(ctx, key, Field(fmt.Sprintf("%s.le_%g", name, boundary), labels), 1)
			}
		}
		pipe.HIncrByFloat(ctx, key, Field(name+".le_inf", labels), 1)
	})
}
func (s *Store) Gauge(ctx context.Context, name string, value any, labels map[string]any) {
	key := s.Prefix + ":gauges"
	pipe := s.Redis.Pipeline()
	pipe.HSet(ctx, key, Field(name, labels), fmt.Sprint(value))
	pipe.Expire(ctx, key, s.Retention)
	_, _ = pipe.Exec(ctx)
}
func (s *Store) Event(ctx context.Context, category, event string, fields map[string]any) {
	payload := map[string]any{"timestamp": time.Now().UnixMilli(), "category": category, "event": event}
	for key, value := range fields {
		if value != nil {
			payload[key] = value
		}
	}
	encoded, _ := json.Marshal(payload)
	key := s.Prefix + ":events"
	pipe := s.Redis.Pipeline()
	pipe.LPush(ctx, key, string(encoded))
	pipe.LTrim(ctx, key, 0, int64(s.EventLimit-1))
	pipe.Expire(ctx, key, s.EventRetention)
	_, _ = pipe.Exec(ctx)
}
func (s *Store) Begin(ctx context.Context, id string) {
	key := s.Prefix + ":inflight"
	pipe := s.Redis.Pipeline()
	pipe.ZAdd(ctx, key, redis.Z{Score: float64(time.Now().Add(5 * time.Minute).Unix()), Member: id})
	pipe.Expire(ctx, key, 10*time.Minute)
	_, _ = pipe.Exec(ctx)
}
func (s *Store) End(ctx context.Context, id string) {
	_ = s.Redis.ZRem(ctx, s.Prefix+":inflight", id).Err()
}
func (s *Store) Inflight(ctx context.Context) int64 {
	key := s.Prefix + ":inflight"
	pipe := s.Redis.Pipeline()
	pipe.ZRemRangeByScore(ctx, key, "-inf", fmt.Sprint(time.Now().Unix()))
	count := pipe.ZCard(ctx, key)
	if _, err := pipe.Exec(ctx); err != nil {
		return 0
	}
	return count.Val()
}
func (s *Store) write(ctx context.Context, action func(redis.Pipeliner, string)) {
	key := s.minuteKey(time.Now())
	pipe := s.Redis.Pipeline()
	action(pipe, key)
	pipe.Expire(ctx, key, s.Retention)
	_, _ = pipe.Exec(ctx)
}
func (s *Store) minuteKey(now time.Time) string {
	return fmt.Sprintf("%s:minute:%d", s.Prefix, now.Unix()/60)
}
func valueOr(value, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}
func intEnv(name string, fallback int) int {
	var value int
	if _, err := fmt.Sscanf(os.Getenv(name), "%d", &value); err == nil && value > 0 {
		return value
	}
	return fallback
}
